### ISSUE TITLE
Support DELIMITER syntax

### ISSUE THREAD

**Issue by Awk34** [mysqljs/mysql#1683]
> Disclaimer: I don't know a whole lot of SQL

So my dev team decided that we wanted a Node script that could drop and re-create all of our stored procedures whenever something changes. The problem was that this library doesn't support the delimiter syntax. I made a kind of hacky way to split statements from my `.sql` (loaded to an in-memory string) file by custom delimiters into an array of separate statements.

Do you guys have a suggested way of working with something like this? How about suggestions for where to start for a PR?

Ex:

```sql
DELIMITER $$
drop procedure `Thing_Insert`$$
CREATE PROCEDURE `Thing_Insert`(
	argA int,
	argB nvarchar(50),
	argC nvarchar(50),
	argD datetime,
	argE bit)
BEGIN

	insert into Thing (A, B, C, D, E)
		select argA, argB, argC, argD, argE;

END$$
DELIMITER ;

...etc
```

```js
/**
 * Separates individual SQL statements, split by a delimiter, into an array of statements
 * @param {String} text
 * @returns {String[]}
 */
function separateSqlCommands(text) {
    let lines = text.split('\n')
        .filter(str => {
            return str.length
                && !/^\n*$/.test(str);
        });

    let delimiter = ';';
    const DELIMITER_REGEX = /DELIMITER (.+)/;

    let statements = [];

    let accumulator = '';

    for(const line of lines) {
        let regexpResult = DELIMITER_REGEX.exec(line);
        if(regexpResult) {
            statements = statements.concat(accumulator.split(delimiter));

            accumulator = '';
            delimiter = regexpResult[1];
        } else {
            accumulator += line + ' ';
        }
    }

    statements = statements.filter(statement => !/^\s*$/.test(statement))
        .map(statement => {
            if(statement[0] === ' ') statement = statement.substr(1);

            return statement.replace(/\t/g, ' ')
                .replace(/ +/g, ' ');
        })
        .map(statement => statement.replace(/ +/g, ' '));

    return statements;
}
```

example output:

```json
[ 'drop procedure `Thing_Insert`',
  'CREATE PROCEDURE `Thing_Insert`( argA int, argB nvarchar(50), argC nvarchar(50), argD datetime, argE bit) BEGIN insert into Thing (A, B, C, D, E) select argA, argB, argC, argD, argE; END' ]
```

**Comment by uldissturms** [comment]
👍 Having the same issue.

**Comment by kpetrow** [comment]
I think you should use DROP PROCEDURE IF EXISTS -- just in case it aint there
and simple batch script:
`mysql -u username -p database_name < file.sql`
no need for node or mysql module

**Comment by Awk34** [comment]
@kpetrow while using the `mysql` CLI directly is always an option, the point here is that one might want to extract out the individual SQL commands and programmatically work with them

**Comment by kpetrow** [comment]
If you are going to work with individual statements then you are going to need to manually parse anyways?  If you are going to slam into db use CLI?

**Comment by dougwilson** [comment]
The hard part about this is that all this module does right now is implement the MySQL protocol, and supporting `DELIMITER` means that the module would have to implement and entire SQL dialect parser to get that added, but if you want to put a PR together, that's certainly good! I know the `separateSqlCommands` is just what you are using now, so I'm sure you're not surprised that it has a few edge cases that would have to be fixed to actually add it here :)

> Do you guys have a suggested way of working with something like this?

I would suggest that if you don't even need to keep it in a `.sql` file that you can load into the `mysql` command line client, then simply store the literal procedure content alone. For example, if you just store your procedure above as
```sql
CREATE PROCEDURE `Thing_Insert`(
	argA int,
	argB nvarchar(50),
	argC nvarchar(50),
	argD datetime,
	argE bit)
BEGIN

	insert into Thing (A, B, C, D, E)
		select argA, argB, argC, argD, argE;

END;
```

Then you can easily use `connection.query` to run that and create the procedure. You can run the drop if exists as a `connection.query` call before that one to get the same effect.

> How about suggestions for where to start for a PR?

If you are looking to add support for `DELIMITER` the first step is to either find a MySQL-dialect parser or build one. Then you can add support for `DELIMITER` and it can be used to slice up the queries to send over the protocol.

**Comment by kpetrow** [comment]
Can you give an example of an edge case?

**Comment by dougwilson** [comment]
Hi @kpetrow , the code above has a lot of issues, the easiest to see are the following:

1. The occurrence of `ADELIMITER $$` will incorrectly change the delimiter to `$$`.
2. The function will incorrectly match a delimiter that is within a string in the SQL.
3. The word `DELIMITER` is not required to be in upper-case, but this will only match upper-case occurrences.
4. The line `foo DELIMITER $$` will incorrectly change the delimiter to `$$`.
6. The line `DELIMITER "$$"` will incorrectly change the delimiter to `"$$"` instead of `$$`.

I can certainly continue on, but it's really just diminishing returns when really the answer is to just fix up the edge cases by implementing it exactly how MySQL implements it.

**Comment by kpetrow** [comment]
I was thinking nested delimeters, but was curious if there was SP, trigger or something that would would need to be implemented in some odd way.  Thanks

**Comment by dougwilson** [comment]
I haven't seen that, but not sure if that's even a thing. That's probably why the best answer is just to implement it exactly how the MySQL client is doing it :) Then we don't have to guess, haha. The protocol was just implemented from the MySQL protocol documentation, but I'm not familiar with any documentation around the delimiter and how it works, exactly.

**Comment by surruk51** [comment]
re-inventing the wheel is not a good idea. It presents maintenance and compatibility issues. There is a solution already. See: https://dev.mysql.com/doc/refman/5.7/en/c-api-multiple-queries.html

**Comment by Awk34** [comment]
@dougwilson Yeah, I was sure there would be plenty of cases where my code wouldn't always work. It was mainly a proof of concept that works well for my use case. Maybe others will find it useful.

I found a SQL parser for JS here: https://github.com/forward/sql-parser. The problem is, at this point, I'm not sure if this should be built into this library, or added as a separate module for working with SQL files using the `DELIMITER` syntax.

**Comment by ben-page** [comment]
I thought I'd clarify a little for anyone coming here looking for DELIMITER support. Most people probably don't need DELIMITER support. The OP only needs it because he's loading external .sql files that were generated for the MySQL CLI.

DELIMITER is needed in the MySQL CLI because it interprets queries containing semicolons as multiple statements. DELIMITER allows you to control the splitting behavior. By default, `mysql` doesn't interpret the query. It just sends it directly to the server. So you can just send queries that include semicolons without a problem.

The OPs queries can be run without DELIMITER support. Just make sure to make a connection with multipleStatements set to false and run the statements one at a time.

```javascript
'use strict';
const mysql = require('mysql');

const config = {
    host: 'localhost',
    user: 'user',
    password: 'password',
    database: 'database',
    multipleStatements: false //false is the default, added for emphasis
};

const dropQuery = 'DROP PROCEDURE IF EXISTS `Thing_Insert`';
const createQuery = `CREATE PROCEDURE \`Thing_Insert\`(
        argA int,
        argB nvarchar(50),
        argC nvarchar(50),
        argD datetime,
        argE bit)
    BEGIN
    
        insert into Thing (A, B, C, D, E)
            select argA, argB, argC, argD, argE;
    
    END`;

const connection = mysql.createConnection(config);
connection.connect(function(err) {
    if (err) {
        console.error('error connecting: ' + err.stack);
        return;
    }
    
    connection.query(dropQuery, function(err2) {
        if (err2) {
            console.error('error querying: ' + err2.stack);
            return;
        }
        
        console.log('drop succeeded');
        
        connection.query(createQuery, function(err3) {
            if (err3) {
                console.error('error querying: ' + err3.stack);
                return;
            }
    
            console.log('create succeeded');
            connection.end();
        });
    });
});
```

**Comment by lagden** [comment]
@ben-page and @dougwilson 

Removing the `DELIMITER` works!!

```javascript
const _sql = sql.toString().replace(/DELIMITER ;?;/gm, '').replace(/;;/gm, ';')
```

Thanks!

**Comment by VeryCrazyDog** [comment]
I spent my weekend and published [a module](https://www.npmjs.com/package/@verycrazydog/mysql-parser) which hope to provide a more standard solution to the missing `DELIMITER` support. I have handled all edge cases pointed out by @dougwilson. It is not a full MySQL language parser but it should be able to complete the work on most of our daily usage scenario.

**Comment by rvalenca** [comment]
Any news on this? Is it still not possible to create a trigger/procedure with multiple statements inside (delimiter case)?

**Comment by rvalenca** [comment]
ok, just figured it out.

If the delimiter lines are removed, somehow it works fine (even though it does not work via a SQL client). So, for example:

```
create or replace trigger trigger1 after insert on table1 for each row
begin
  declare _var1 int(11);
  select a into _var1 from table1;
end
```

OBS.: I did the test using the 'mariadb' lib, but I suppose it will be the same for 'mysql' as well.

**Comment by aamirpinger** [comment]
> ok, just figured it out.
> 
> If the delimiter lines are removed, somehow it works fine (even though it does not work via a SQL client). So, for example:
> 
> ```
> create or replace trigger trigger1 after insert on table1 for each row
> begin
>   declare _var1 int(11);
>   select a into _var1 from table1;
> end
> ```

Thank you! +1 for above!
